GoogleForCreators/web-stories-wp
# Media: Do not store transient data on resource objects

## Context



## Summary
Remove the `local`, `isMuting`, `isTrimming` and `isTranscoding` properties from the resource object stored in elements. This means, a large refactor, meaning all the references to this properties, in code and in tests. Use the media queue hook to work out if an resource is transcoding or local. 

There are two kinds of resources lookups required. By ID or by original id. For muting/transcoding/trimming resources, there are cases where need to lookup by id ( like in the media library ) and cases where we need to lookup state by the original id. This is because, while a resource is processing, it doesn't have an id yet, it uses a temp id. So we have to lookup via the old resources id. 

I have fixed all tests and added a couple of new ones for the new functionality found in the reducer. 

## Relevant Technical Choices



## To-do



## User-facing changes



## Testing Instructions






This PR can be tested by following these steps:

This change effects all parts of uploading assets, ALL upload interactions will need to be tests. Including. 

- Upload of Image, video and gifs. 
- Muting of videos. 
- Trimming of videos and trimming a trimmed video. 
- Transcoding videos. 
- Uploading via the WP media dialog. 
- Dragging into canvas. 
- Pasting image into canvas
- Dragging animated gif onto canvas and convert to mp4. 

All video uploads, should include a time stamp, is muted icon and poster image. 

### Upload of Image.
1. Open editor. 
2. Drag a png/jpeg file into canvas / media library. 
3. See progress bar appear on new media item added to library and save button disabled.
4. Progress bar finishes. 
5. Save button re-enabled 

### Upload of Video.
1. Open editor. 
2. Drag a mp4/webm file into canvas / media library. 
3. See progress bar appear on new media item added to library and save button disabled. You should be able to hover over save button / new item in media library, saying that transcoding is taking place.  Trimming / muting options should be hidden. 
4. Progress bar finishes. 
5. Save button re-enabled / Trimming / muting buttons should appear

### Upload of animated gif.
1. Open editor. 
2. Drag a animated fig file into canvas / media library. 
3. See progress bar appear on new media item added to library and save button disabled. You should be able to hover over save button / new item in media library, saying that transcoding is taking place. 
4. Progress bar finishes. 
5. Save button re-enabled 

### Muting a video. 
1. Open editor.
2. Insert a video ( that has sound ) into canvas. 
3. In documents sidebar, under video options click mute video. 
4. See progress bar appear on new media item added to library and save button disabled. You should be able to hover over save button / new item in media library, saying that transcoding is taking place. 
5. Progress bar finishes. 
6. Save button re-enabled 

### Trimming video
1. Open editor.
2. Insert a video into canvas. 
3. In documents sidebar, under video options click trim. 
4. Trimming video to 10 seconds. 
5. Click trim to start trimming process. 
6. See progress bar appear on new media item added to library and save button disabled. You should be able to hover over save button / new item in media library, saying that transcoding is taking place. 
7. Progress bar finishes. 
8. Save button re-enabled. 

### Retrimming video
1. Follow steps in trimming video. 
2. On video video, retrim video, following steps in trimming video.

### Pasting file
1. Copy jpg/mp4/gif into clip from desktop / folder. 
2. Open editor. 
3. Paste file into canvas. 
4. See progress bar appear on new media item added to library and save button disabled.
5. Progress bar finishes. 
6. Save button re-enabled 

### Upload via the WP media dialog. 
1. Open editor. 
2. Click on upload button. 
3. See WP media dialog open. 
4. Upload jpg/animated gif/video into WP media dialog. 
5. Insert new asset into page. 
6. See dialog close. 
7. See new asset in media library and element in canvas. 

### Transcode existing video. 
1. Go to Media > Upload
2. Upload 1080p / 4K mp4 video. 
3. Open editor. 
4. Insert 1080p / 4K mp4 video into canvas.
5. Add title to story. 
6. Publish story. 
7. Open checklist. 
8. See option to opertize video. 
9. Click opertize video. 
10. See progress bar appear on new media item added to library and save button disabled. You should be able to hover over save button / new item in media library, saying that transcoding is taking place. 
11. Progress bar finishes. 
12. Save button re-enabled. 

## Reviews

### Does this PR have a security-related impact?



### Does this PR change what data or activity we track or use?



### Does this PR have a legal-related impact?



## Checklist




---



Fixes #9577
